src/shapes.cpp

```diff
@@ -564,6 +556,70 @@ void Mesh::computeVertexNormals()
   }
 }
 
+void Mesh::computeWeightedVertexNormals()
+{
+  if (!triangle_normals)
+    computeTriangleNormals();
+  if (vertex_count && !vertex_normals)
+    vertex_normals = new double[vertex_count * 3];
+  EigenSTL::vector_Vector3d avg_normals(vertex_count, Eigen::Vector3d(0, 0, 0));
+
+  for (unsigned int tIdx = 0; tIdx < triangle_count; ++tIdx)
+  {
+    unsigned int tIdx3 = 3 * tIdx;
+    unsigned int tIdx3_1 = tIdx3 + 1;
+    unsigned int tIdx3_2 = tIdx3 + 2;
+
+    unsigned int v1 = triangles[tIdx3];
+    unsigned int v2 = triangles[tIdx3_1];
+    unsigned int v3 = triangles[tIdx3_2];
+
+    // Get angles for each vertex at this triangle
+    Eigen::Vector3d p1{ vertices[3 * v1], vertices[3 * v1 + 1], vertices[3 * v1 + 2] };
+    Eigen::Vector3d p2{ vertices[3 * v2], vertices[3 * v2 + 1], vertices[3 * v2 + 2] };
+    Eigen::Vector3d p3{ vertices[3 * v3], vertices[3 * v3 + 1], vertices[3 * v3 + 2] };
```

Comment:
Consider using `Eigen::Map<Eigen::Vector3d>(vertices + 3*vi, 3)` to avoid copying the data.